### DISCUSSION TITLE
Using brew in a CI/CD pipeline

### DISCUSSION THREAD

**Discussion by davexre** [Homebrew/discussions#1962] [Everyday usage] [answered]
We've been trying to make use of brew to build/maintain several packages in a GitLab CI/CD environment, and we're running into some issues. It seems like maybe we're doing something that brew isn't really intended for (based on some of the observed "under the covers" behaviors) - or perhaps, we're just doing it wrong. 

Briefly - we customize and build several packages locally for use in our product builds. We do this on Mac OS, Windows, and linux. We attempt to use appropriate package management strategies on each platform (so, nuget on Windows, homebrew on Mac OS, and RPM on linux). For Mac, we provide these packages as Bottles via a Tap for normal use. This works fine. Where we run into trouble is when we need to maintain or update one of those packages. 

The way this normally works in a GitLab CI/CD pipeline is that you simply pull a clone of the working branch rather than master, and run the pipeline against that clone. With a build using brew, we have to tap the individual project, manipulate the tap dir via git to get it on the correct branch, then build the bottle, build a bottle description JSON that points at our artifactory server and merge that JSON to the formula. Roughly, this is:

```
brew tap <project-name> <project URL on GitLab>
git -C <tap-dir> fetch
git -C <tap-dir> checkout -f <build branch>
git -C <tap-dir> reset --hard origin/<build branch> --
brew install --build-bottle <formula path>
brew bottle <formula path> --root-url=<artifactory url> --json --no-rebuild
brew bottle --merge *.bottle.json --write --no-commit
```

**_The big question is - are we doing this correctly? Is there a better way to have brew play within a CI/CD pipeline like this?_**  Any pointers to docs we may have missed, or techniques that work better, etc, would be appreciated - as well as any other discussion, obviously. I know we can prevent brew from auto-updating, so we can ensure that future brew changes don't unwittingly disrupt our pipeline. That's not an optimal solution, but we'll use it, if we need to.

When I opened this topic, we were also having an acute issue that some help from @carlocab allowed us to find - the rest of the post that follows was related to that issue, and isn't relevant to the question above. If I need to open a different topic for this discussion, let me know? 

--------

These packages are changed and built infrequently - the last time was two months ago. Something changed in the last two months, and now it appears that the first brew command (the `brew install` command) is failing due to a `GitHub::API::ValidationFailedError`. (I'm still working on confirming that's specifically the culprit - but the build of our code seems to succeed, with the failure happening sometime after as part of brew performing the `system` command on line 13 of the formula - formula provided below). 

Additional info:

The formula.rb file (obfuscated company specific info below, since this is a publicly readable post):
```
class FormulaClassName < Formula
    desc "description"
    url "gitlab/project/url", :using => :git, :branch => ENV["HOMEBREW_BRANCH"]
    version "1.1.0-1"
    depends_on "company-common-tap" => :build
    depends_on "mbedtls" => :build
    depends_on "spdlog" => :build
    depends_on "libcurl" => :build
    depends_on "libzip" => :build
    depends_on "rpm" => :build
    depends_on "cmake" => :build
  def install
    system "export JAVA_HOME=${HOMEBREW_JAVA_HOME} && make build"
    bin.install "build/executable"
    bin.install "tests/path/related.jar"
    end
end
```

And the full text of the current error from the CI/CD pipeline (I've obfuscated company specific info by hand below, since this is a publicly readable post):

```
$ brew tap company/${CI_PROJECT_NAME} ${CI_REPOSITORY_URL}
==> Tapping project-tap-name
Cloning into '/usr/local/Homebrew/Library/Taps/company/homebrew-project'...
Tapped 1 formula (29 files, 75.6KB).
$ git -C ${TAP_DIR} fetch
$ git -C ${TAP_DIR} checkout -f ${CI_BUILD_REF_NAME}
Switched to a new branch 'new-tests'
Branch 'new-tests' set up to track remote branch 'new-tests' from 'origin'.
$ git -C ${TAP_DIR} reset --hard origin/${CI_BUILD_REF_NAME} --
HEAD is now at dd6d0c3 commit text
$ brew install --build-bottle company/${CI_PROJECT_NAME}/xx${CI_PROJECT_NAME}
==> Installing project-name from gitlab-path
==> Cloning gitlab-server/appliance/platform/java/project-name.git
Updating /Users/administrator/Library/Caches/Homebrew/project-name--git
==> Checking out branch new-tests
Already on 'new-tests'
Your branch is up to date with 'origin/new-tests'.
HEAD is now at dd6d0c3 commit text
Warning: Your Xcode (11.4.1) is outdated.
Please update to Xcode 12.4 (or delete it).
Xcode can be updated from the App Store.
Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force
If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install
Alternatively, manually download them from:
  https://developer.apple.com/download/more/.
You should download the Command Line Tools for Xcode 12.4.
==> export JAVA_HOME=${HOMEBREW_JAVA_HOME} && make build
Last 15 lines from /Users/administrator/Library/Logs/Homebrew/project-name/01.export:
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/project-20210806-38007-10hg2jf/build
[ 80%] Building CXX object CMakeFiles/project.dir/src/file1.cpp.o
[ 80%] Building CXX object CMakeFiles/project.dir/src/file2.cpp.o
[ 80%] Building CXX object CMakeFiles/project.dir/src/file3.cpp.o
[ 80%] Building CXX object CMakeFiles/project.dir/src/main.cpp.o
[100%] Linking CXX executable project
[100%] Built target project
If reporting this issue please do so to (not Homebrew/brew or Homebrew/core):
  <obfuscated project tap name>
/usr/local/Homebrew/Library/Homebrew/utils/github/api.rb:304:in `raise_error': Validation Failed: [{"message"=>"The listed users and repositories cannot be searched either because the resources do not exist or you do not have permission to view them.", "resource"=>"Search", "field"=>"q", "code"=>"invalid"}] (GitHub::API::ValidationFailedError)
	from /usr/local/Homebrew/Library/Homebrew/utils/github/api.rb:234:in `open_rest'
	from /usr/local/Homebrew/Library/Homebrew/utils/github.rb:160:in `search'
	from /usr/local/Homebrew/Library/Homebrew/utils/github.rb:30:in `search_issues'
	from /usr/local/Homebrew/Library/Homebrew/utils/github.rb:61:in `issues_for_formula'
	from /usr/local/Homebrew/Library/Homebrew/exceptions.rb:459:in `fetch_issues'
	from /usr/local/Homebrew/Library/Homebrew/exceptions.rb:455:in `issues'
	from /usr/local/Homebrew/Library/Homebrew/exceptions.rb:509:in `dump'
	from /usr/local/Homebrew/Library/Homebrew/brew.rb:176:in `rescue in <main>'
	from /usr/local/Homebrew/Library/Homebrew/brew.rb:164:in `<main>'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2249:in `block in system': Failed executing: export JAVA_HOME=${HOMEBREW_JAVA_HOME} && make build (BuildError)
	from /usr/local/Homebrew/Library/Homebrew/formula.rb:2185:in `open'
	from /usr/local/Homebrew/Library/Homebrew/formula.rb:2185:in `system'
	from /usr/local/Homebrew/Library/Taps/company/homebrew-project/Formula/our-formula.rb:13:in `install'
	from /usr/local/Homebrew/Library/Homebrew/build.rb:178:in `block (2 levels) in install'
	from /usr/local/Homebrew/Library/Homebrew/formula.rb:1281:in `block in brew'
	from /usr/local/Homebrew/Library/Homebrew/formula.rb:2415:in `block (2 levels) in stage'
	from /usr/local/Homebrew/Library/Homebrew/utils.rb:550:in `with_env'
	from /usr/local/Homebrew/Library/Homebrew/formula.rb:2414:in `block in stage'
	from /usr/local/Homebrew/Library/Homebrew/resource.rb:121:in `block (2 levels) in unpack'
	from /usr/local/Homebrew/Library/Homebrew/download_strategy.rb:110:in `chdir'
	from /usr/local/Homebrew/Library/Homebrew/download_strategy.rb:102:in `stage'
	from /usr/local/Homebrew/Library/Homebrew/resource.rb:117:in `block in unpack'
	from /usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `block in run'
	from /usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `chdir'
	from /usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `run'
	from /usr/local/Homebrew/Library/Homebrew/resource.rb:199:in `mktemp'
	from /usr/local/Homebrew/Library/Homebrew/resource.rb:116:in `unpack'
	from /usr/local/Homebrew/Library/Homebrew/resource.rb:91:in `stage'
	from /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/lib/ruby/2.6.0/forwardable.rb:230:in `stage'
	from /usr/local/Homebrew/Library/Homebrew/formula.rb:2394:in `stage'
	from /usr/local/Homebrew/Library/Homebrew/formula.rb:1274:in `brew'
	from /usr/local/Homebrew/Library/Homebrew/build.rb:138:in `block in install'
	from /usr/local/Homebrew/Library/Homebrew/utils.rb:550:in `with_env'
	from /usr/local/Homebrew/Library/Homebrew/build.rb:133:in `install'
	from /usr/local/Homebrew/Library/Homebrew/build.rb:229:in `<main>'
Cleaning up file based variables
00:00
ERROR: Job failed: exit status 1
```

**Comment by carlocab** [comment]
I don't think your failure is from the `GitHub::API::ValidationFailedError`. I think it happened in this step:
```ruby
system "export JAVA_HOME=${HOMEBREW_JAVA_HOME} && make build"
```
To test this, do something trivial in your install method (e.g. `touch prefix/"foo"`), and then see if you still get the same failure. `brew` only queries the GitHub API after a build failure, and not randomly inside an `install` method. This means that what actually happened is that your build failed for some reason and then `brew` hit another failure when it tried to query GitHub to produce error messages about your failure.

It might help to change that first command in your install method to
```ruby
ENV["JAVA_HOME"] = ENV["HOMEBREW_JAVA_HOME"]
system "make", "build"
```

**Reply by davexre** [reply]
Hi - thanks for that info. I'll perform some additional steps and see if I can shake out the root on that failure. Definitely helps to know that brew isn't calling the GitHub API without an actual failure prior to it. 

Would you have any thoughts on the general CI/CD issue?

**Reply by davexre** [reply]
Ok, that clue allowed us to find root cause on the latest issue. The "something" that changed was part of our build process - maven was pulled into the mix, and the code didn't have a full path to the mvn executable. So, many thanks there - that's a good clue.

I'd still like to leave this open for discussion on the general CI/CD topic (and will edit my original post a bit to reflect that)?

**Comment by davexre** [comment, marked as answer]
Just a friendly bump on this one - we're still hoping to get feedback on general use of homebrew in a CI/CD environment?

**Reply by SMillerDev** [reply]
Reading the post again it seems like you're trying to make everything happen in a single build, while I think for most projects that trigger tap updates they have that separated. You have the main build process that runs `brew bump-formula-pr`, which will send a pull request to build the update to the tap.

I'd say by trying to keep the manipulation of the second repo (the tap) in the CI of the first you're making everything harder for yourself.

**Reply by davexre** [reply]
I may not be understanding correctly - my understanding was that `brew bump-formula-pr` was only useful for projects that published a formula to the homebrew GItHub repo?

Really, all we're trying to do is checkout a specific branch from the internal repo, and build that branch so that we can run specific tests against it to make sure it's working. Internally, we follow a merge request model in GitLab. When an MR is filed against a project within the repo, GitLab automatically runs the branch indicated in the MR through a CI pipeline to test it - in order to successfully build a formula from a branch that's not the default/master branch, we've had to adopt those steps. This is all happening out of the same repo that's used as the tap, not a second repo.

**Reply by SMillerDev** [reply]
> my understanding was that brew bump-formula-pr was only useful for projects that published a formula to the homebrew GItHub repo?

No, it can bump any tap. Although you're right that it works better with github.

> This is all happening out of the same repo that's used as the tap, not a second repo.

Wait, your software lives in the same repo as the tap? That's really unusual and likely contributes to the confusion.

Overall it seems like you're using brew as your build tool. It's really not meant for that. It's meant for package management. You can build your software using make/gradle/meson/whatever and just package it with brew. Any other configuration is asking for compatibility problems IYAM.

**Reply by davexre** [reply]
> Wait, your software lives in the same repo as the tap? That's really unusual and likely contributes to the confusion.
> 
> Overall it seems like you're using brew as your build tool. It's really not meant for that. It's meant for package management. You can build your software using make/gradle/meson/whatever and just package it with brew. Any other configuration is asking for compatibility problems IYAM.

Yeah, that seems accurate. The Formula is calling make as part of installing it. I suspected we might be doing something that was at odds with how brew wanted to work. We'll investigate how to alter our workflow to be a little more in line with solely packaging things with brew, and maybe this issue will just sort of go away for us. 

Thanks for the help and explanation!
